include/revng/Model/Binary.h

```diff
@@ -118,6 +119,13 @@ class model::Binary : public model::generated::Binary {
 
   model::TypePath recordNewType(UpcastablePointer<Type> &&T);
 
+  template<derived_from<model::Type> NewType>
+  std::pair<NewType &, model::TypePath> makeType() {
+    model::UpcastableType Result = model::UpcastableType::make<NewType>();
+    model::TypePath ResultPath = recordNewType(std::move(Result));
+    return { *llvm::dyn_cast<NewType>(ResultPath.get()), ResultPath };
```

Comment:
Wait, maybe I'm misunderstanding something, but it's it guaranteed to never be null?

The object pointer to which it is casting and dereferencing was created right before it. Unless there is some nasty race condition, I cannot imagine the situation where this could end up badly. Am I wrong?

Reply:
You're right, but still `dyn_cast` is a footgun. You're not happy if `dyn_cast` ever returns `nullptr`, so go for `llvm::cast`.

Reply:
Done: https://github.com/revng/revng/pull/232/commits/d2253401ee0331afb7ad79422f9c4889aed363cd.